pkg/controller/notification/snstopic/controller.go

```diff
@@ -0,0 +1,221 @@
+	errClient                = "cannot create a new SNSTopic client"
+	errUnexpectedObject      = "The managed resource is not a SNSTopic resource"
+	errList                  = "failed to list SNS Topics"
+	errGetTopic              = "failed to get SNS Topic"
+	errListTopic             = "failed to list SNS Topics"
+	errGetTopicAttr          = "failed to get SNS Topic Attribute"
+	errCreate                = "failed to create the SNS Topic"
+	errDelete                = "failed to delete the SNS Topic"
+	errUpdate                = "failed to update the SNS Topic"
+)
+
+// SetupSNSTopic adds a controller that reconciles SNSTopic.
+func SetupSNSTopic(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.SNSTopicGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.SNSTopic{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.SNSTopicGroupVersionKind),
+			managed.WithExternalConnecter(&connector{
+				kube:        mgr.GetClient(),
+				newClientFn: sns.NewTopicClient,
+				awsConfigFn: utils.RetrieveAwsConfigFromProvider,
+			}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithInitializers(),
+			managed.WithConnectionPublishers(),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	kube        client.Client
+	newClientFn func(*aws.Config) (sns.TopicClient, error)
+	awsConfigFn func(context.Context, client.Reader, *corev1.ObjectReference) (*aws.Config, error)
+}
+
+func (conn *connector) Connect(ctx context.Context, mgd resource.Managed) (managed.ExternalClient, error) {
+
+	cr, ok := mgd.(*v1alpha1.SNSTopic)
+	if !ok {
+		return nil, errors.New(errUnexpectedObject)
+	}
+
+	awsconfig, err := conn.awsConfigFn(ctx, conn.kube, cr.Spec.ProviderReference)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := conn.newClientFn(awsconfig)
+	if err != nil {
+		return nil, errors.Wrap(err, errClient)
+	}
+	return &external{c, conn.kube}, nil
+}
+
+type external struct {
+	client snsclient.TopicClient
+	kube   client.Client
+}
+
+func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mgd.(*v1alpha1.SNSTopic)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+	// AWS SNS resources are uniquely identified by an ARN that is returned
+	// on create time; we can't tell whether they exist unless we have recorded
+	// their ARN.
+	if !awsarn.IsARN(meta.GetExternalName(cr)) {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+
+	// Fetch the list of Topics
+	topicList, err := e.client.ListTopicsRequest(&awssns.ListTopicsInput{}).Send(ctx)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(sns.IsTopicNotFound, err), errList)
+	}
```

Comment:
Similar to my comment about subscriptions, is there a value in listing topics? it seems like we could just go straight to calling `GetTopicAttributesRequest` and determine whether the topic exists based on whether that API call returns a HTTP 404.

I see that we select a listed topic and use that for `LateInitializeTopic`, but per my comment on that function I don't think we need to.